bionic/cache_api.py

```diff
@@ -0,0 +1,108 @@
+    object. Use ``get_entries`` to iterate through the set of cache entries.
+    """
+
+    def __init__(self, flow):
+        self._flow = flow
+
+    def get_entries(self):
+        """
+        Returns a sequence of :class:`CacheEntry <bionic.cache_api.CacheEntry>`
+        objects, one for each artifact in Bionic's persistent cache.
+        """
+
+        # These private accesses are a bit gross, but I'm not sure it's worth adding
+        # more layers of APIs to avoid them.
+        deriver = self._flow._deriver
+        deriver.get_ready()
+        persistent_cache = deriver._bootstrap.persistent_cache
+
+        stores = [
+            store
+            for store in [persistent_cache._local_store, persistent_cache._cloud_store]
+            if store is not None
+        ]
+
+        return (
+            CacheEntry(cache=self, inv_item=inv_item,)
+            for store in stores
+            for inv_item in store.inventory.list_items()
+        )
+
+
+def path_from_url_if_file(url):
+    """Converts a URL into a file path if it is a file URL; otherwise returns None."""
+
+    if is_file_url(url):
+        return path_from_url(url)
+    else:
+        return None
+
+
+@total_ordering
+class CacheEntry:
+    """
+    Represents an artifact in Bionic's persistent cache.
+
+    Has the following fields:
+
+    - ``tier``: "local" or "cloud", depending on which tier of the cache the artifact
+      is in.
+    - ``entity``: the name of the cached entity, or ``None`` if the artifact is does
+      not correspond to an entity
+    - ``artifact_url``: a URL to the cached artifact file or blob
+    - ``metadata_url``: a URL to the metadata file or blob describing the artifact
+    - ``artifact_path``: a Path object locating the artifact file, if it's a file
+    - ``metadata_path``: a Path object locating the metadata file, if it's a file
+    """
+
+    def __init__(self, cache, inv_item):
+        self._cache = cache
+        self._comparison_key = inv_item.abs_metadata_url
+
+        self.tier = inv_item.inventory.tier
+        self.artifact_url = inv_item.abs_artifact_url
+        self.metadata_url = inv_item.abs_metadata_url
+        self._descriptor = inv_item.descriptor
+        self._inventory = inv_item.inventory
+
+    @property
+    def entity(self):
+        try:
+            return entity_dnode_from_descriptor(self._descriptor).to_entity_name()
+        except ValueError:
+            return None
+
+    @property
+    def artifact_path(self):
+        return path_from_url_if_file(self.artifact_url)
+
+    @property
+    def metadata_path(self):
+        return path_from_url_if_file(self.metadata_url)
+
+    def __hash__(self):
+        return hash(self._comparison_key)
+
+    def __eq__(self, other):
+        if not isinstance(other, CacheEntry):
+            return False
+        return self._comparison_key == other._comparison_key
+
+    def __lt__(self, other):
```

Comment:
Do you not need to check that `other` is an instance of `CacheEntry` in `lt`?

Reply:
I think we have to throw an exception either way, but yeah, we can make that exception a `TypeError`.